diff --git a/meta-phosphor/classes/obmc-phosphor-image_types_uboot.bbclass b/meta-phosphor/classes/obmc-phosphor-image_types_uboot.bbclass
index 01f91c0..a5c0222 100644
--- a/meta-phosphor/classes/obmc-phosphor-image_types_uboot.bbclass
+++ b/meta-phosphor/classes/obmc-phosphor-image_types_uboot.bbclass
@@ -20,7 +20,8 @@ FLASH_IMAGE_LINK ?= "flash-${MACHINE}"
 FLASH_KERNEL_IMAGETYPE ?= "cuImage"
 
 FLASH_UBOOT_OFFSET ?= "0"
-FLASH_KERNEL_OFFSET ?= "512"
+FLASH_DTB_OFFSET ?= "512"
+FLASH_KERNEL_OFFSET ?= "576"
 FLASH_INITRD_OFFSET ?= "3072"
 FLASH_ROFS_OFFSET ?= "4864"
 FLASH_RWFS_OFFSET ?= "28672"
@@ -56,6 +57,8 @@ do_generate_flash() {
        rwfs="rwfs.${OVERLAY_BASETYPE}"
        rofsimg=rofs.${IMAGE_BASETYPE}.cpio
        netimg=initramfs-netboot.cpio
+#       dtb="${FLASH_KERNEL_IMAGETYPE}-${KERNEL_DEVICETREE}"
+       dtb="zImage-${KERNEL_DEVICETREE}"
 
        if [ ! -f $ddir/$kernel ]; then
               bbfatal "Kernel file ${ddir}/${kernel} does not exist"
@@ -82,6 +85,7 @@ do_generate_flash() {
        rm -rf $dst
        mk_nor_image ${dst} ${FLASH_SIZE}
        dd if=${ddir}/${uboot} of=${dst} bs=1k conv=notrunc seek=${FLASH_UBOOT_OFFSET}
+       dd if=${ddir}/${dtb} of=${dst} bs=1k conv=notrunc seek=${FLASH_DTB_OFFSET}
        dd if=${ddir}/${kernel} of=${dst} bs=1k conv=notrunc seek=${FLASH_KERNEL_OFFSET}
        dd if=${ddir}/${uinitrd} of=${dst} bs=1k conv=notrunc seek=${FLASH_INITRD_OFFSET}
        dd if=${ddir}/${rootfs} of=${dst} bs=1k conv=notrunc seek=${FLASH_ROFS_OFFSET}
