diff -Naurp source/drivers/char/ipmi.orig/Kconfig source/drivers/char/ipmi/Kconfig
--- source/drivers/char/ipmi.orig/Kconfig	2016-11-22 18:28:30.000000000 +0300
+++ source/drivers/char/ipmi/Kconfig	2016-11-23 06:35:46.943922989 +0300
@@ -70,6 +70,13 @@ config IPMI_SSIF
 	 have a driver that must be accessed over an I2C bus instead of a
 	 standard interface.  This module requires I2C support.
 
+config IPMI_I2C
+       tristate 'IPMB I2C handler'
+       select I2C
+       help
+         Provides a driver for a IPMB interface to the remote BMC.
+	 This module requires I2C support.
+
 config IPMI_POWERNV
        depends on PPC_POWERNV
        tristate 'POWERNV (OPAL firmware) IPMI interface'
diff -Naurp source/drivers/char/ipmi.orig/Makefile source/drivers/char/ipmi/Makefile
--- source/drivers/char/ipmi.orig/Makefile	2016-11-22 18:28:30.000000000 +0300
+++ source/drivers/char/ipmi/Makefile	2016-11-23 06:36:08.799581712 +0300
@@ -8,6 +8,7 @@ obj-$(CONFIG_IPMI_HANDLER) += ipmi_msgha
 obj-$(CONFIG_IPMI_DEVICE_INTERFACE) += ipmi_devintf.o
 obj-$(CONFIG_IPMI_SI) += ipmi_si.o
 obj-$(CONFIG_IPMI_SSIF) += ipmi_ssif.o
+obj-$(CONFIG_IPMI_I2C) += ipmi_i2c.o
 obj-$(CONFIG_IPMI_POWERNV) += ipmi_powernv.o
 obj-$(CONFIG_IPMI_WATCHDOG) += ipmi_watchdog.o
 obj-$(CONFIG_IPMI_POWEROFF) += ipmi_poweroff.o
