--- kernel-source/drivers/net/ethernet/faraday/ftgmac100.c.orig	2017-08-01 19:44:28.236690893 +0300
+++ kernel-source/drivers/net/ethernet/faraday/ftgmac100.c	2017-08-01 19:44:10.928549932 +0300
@@ -221,7 +221,8 @@
 				 FTGMAC100_MACCR_CRC_APD	| \
 				 FTGMAC100_MACCR_PHY_LINK_LEVEL | \
 				 FTGMAC100_MACCR_RX_RUNT	| \
-				 FTGMAC100_MACCR_RX_BROADPKT)
+				 FTGMAC100_MACCR_RX_BROADPKT    | \
+                                 FTGMAC100_MACCR_RX_MULTIPKT)
 
 static void ftgmac100_start_hw(struct ftgmac100 *priv, int speed)
 {
