--- git/include/configs/ast-common.h.orig	2016-11-18 20:26:40.000000000 +0300
+++ git/include/configs/ast-common.h	2016-11-19 08:00:57.696922911 +0300
@@ -84,7 +84,7 @@
 #define CONFIG_SYS_MAXARGS		16
 #define CONFIG_SYS_BARGSIZE		CONFIG_SYS_CBSIZE
 
-#define CONFIG_BOOTARGS			"console=ttyS4,115200n8 root=/dev/ram rw"
+#define CONFIG_BOOTARGS			"console=ttyS4,115200n8 root=/dev/ram rw earlyprintk"
 
 #ifdef CONFIG_AST_SPI_NOR    /* AST SPI NOR Flash */
 #define CONFIG_SYS_MAX_FLASH_BANKS 	(CONFIG_FMC_CS)
@@ -94,9 +94,9 @@
 #endif
 
 #define CONFIG_ENV_OFFSET		0x60000	/* environment starts here  */
-#define CONFIG_ENV_SIZE			0x20000	/* Total Size of Environment Sector */
+#define CONFIG_ENV_SIZE			0x20000	/* Total Size of Environment Sector */
 
-#define CONFIG_BOOTCOMMAND	"bootm 20080000 20300000"
+#define CONFIG_BOOTCOMMAND	"mw.l 1e78502c 00000092 1; bootm 20090000 20300000 20080000"
 #define CONFIG_ENV_OVERWRITE
 
 #define CONFIG_SKIP_LOWLEVEL_INIT 1
