--- git/include/configs/ast-common.h.orig	2016-11-18 20:26:40.000000000 +0300
+++ git/include/configs/ast-common.h	2016-11-19 08:00:57.696922911 +0300
@@ -84,7 +84,7 @@
 #define CONFIG_SYS_MAXARGS		16
 #define CONFIG_SYS_BARGSIZE		CONFIG_SYS_CBSIZE
 
-#define CONFIG_BOOTARGS			"console=ttyS4,115200n8 root=/dev/ram rw"
+#define CONFIG_BOOTARGS			"console=ttyS4,115200n8 root=/dev/ram rw earlyprintk"
 
 #ifdef CONFIG_AST_SPI_NOR    /* AST SPI NOR Flash */
 #define CONFIG_SYS_MAX_FLASH_BANKS 	(CONFIG_FMC_CS)
@@ -99,7 +99,8 @@
 #define CONFIG_BOOTCOMMAND	\
 	"fdt addr 20080000; " \
 	"if fdt get value ramdisk_conf /configurations/conf@1 ramdisk; then " \
-	"    bootm 20080000; else bootm 20080000 20300000; " \
+	"    mw.l 1e78502c 00000092 1; bootm 20080000; else " \
+	"    mw.l 1e78502c 00000092 1; bootm 20090000 20300000 20080000; " \
 	"fi"
 #define CONFIG_ENV_OVERWRITE
 
